subprojects/tooling-native/src/crossVersionTest/groovy/org/gradle/language/cpp/tooling/r410/CppModelCrossVersionSpec.groovy

```diff
@@ -125,6 +127,139 @@ class CppModelCrossVersionSpec extends ToolingApiSpecification {
         project.testComponent == null
     }
 
+    @ToolingApiVersion(">=5.2")
+    @TargetGradleVersion(">=5.2")
+    def "can query model when root project applies C++ application plugin with multiple architecture"() {
+        settingsFile << """
+            rootProject.name = 'app'
+        """
+        buildFile << """
+            apply plugin: 'cpp-application'
+
+            application {
+                targetMachines = [machines.${currentHostOperatingSystemFamilyDsl}.x86, machines.${currentHostOperatingSystemFamilyDsl}.x86_64]
+            }
+        """
+        def headerDir = file('src/main/headers')
+        def src1 = file('src/main/cpp/app.cpp').createFile()
+        def src2 = file('src/main/cpp/app-impl.cpp').createFile()
+
+        when:
+        def project = withConnection { connection -> connection.getModel(CppProject.class) }
+
+        then:
+        project.projectIdentifier.projectPath == ':'
+        project.projectIdentifier.buildIdentifier.rootDir == projectDir
+
+        project.mainComponent instanceof CppApplication
+        project.mainComponent.name == 'main'
+        project.mainComponent.baseName == 'app'
+
+        project.mainComponent.binaries.size() == 4
+
+        def debugX86Binary = project.mainComponent.binaries[0]
+        debugX86Binary instanceof CppExecutable
+        debugX86Binary.name == 'mainDebugX86'
+        debugX86Binary.variantName == 'debugX86'
+        debugX86Binary.baseName == 'app'
+        debugX86Binary.compilationDetails.sources.sourceFile as Set == [src1, src2] as Set
+        debugX86Binary.compilationDetails.sources.objectFile.each { assert it != null }
+        debugX86Binary.compilationDetails.headerDirs == [headerDir] as Set
+        debugX86Binary.compilationDetails.frameworkSearchPaths.empty
+        !debugX86Binary.compilationDetails.systemHeaderSearchPaths.empty
+        debugX86Binary.compilationDetails.userHeaderSearchPaths == [headerDir]
+        debugX86Binary.compilationDetails.macroDefines.empty
+        debugX86Binary.compilationDetails.macroUndefines.empty
+        debugX86Binary.compilationDetails.additionalArgs.empty
+        debugX86Binary.compilationDetails.compilerExecutable == toolchain.cppCompiler
```

Comment:
This doesn't quite work for x86 on Windows because there are two different compilers used (one for building x86 and one for x86-64).  I think to do this right, we'd need to make the AvailableToolChain fixture smarter to probe based on the target machine.  I don't think we should do that right now.

I think I would relax this check to only assert that the compilerExecutable isn't null for these multi-target tests.

Reply:
I relaxed the check by just comparing the last part of the compiler `File`.